builtin/common/serialize.lua

```diff
@@ -1,205 +1,232 @@
+local function count_objects(value)
+	local counts = {}
+	if value == nil then
+		-- Early return for nil; tables can't contain nil
+		return counts
+	end
+	local function count_values(val)
+		local type_ = type(val)
+		if type_ == "boolean" or type_ == "number" then
 			return
 		end
-		if seen[x] == 1 then
-			seen[x] = 2
-		elseif seen[x] ~= 2 then
-			seen[x] = 1
-		end
-
-		if tp == "table" then
-			nested[x] = true
-			for k, v in pairs(x) do
-				if nested[k] or nested[v] then
-					mark_nest_point(x, k, v)
-				else
-					mark_multiple_occurences(k)
-					mark_multiple_occurences(v)
+		local count = counts[val]
+		counts[val] = (count or 0) + 1
+		if type_ == "table" then
+			if not count then
+				for k, v in pairs(val) do
+					count_values(k)
+					count_values(v)
 				end
 			end
-			nested[x] = nil
+		elseif type_ ~= "string" and type_ ~= "function" then
+			error("unsupported type: " .. type_)
 		end
 	end
+	count_values(value)
+	return counts
+end
 
-	local dumped     = {}  -- object->varname set
-	local local_defs = {}  -- Dumped local definitions as source code lines
+-- Build a "set" of Lua keywords. These can't be used as short key names.
+-- See https://www.lua.org/manual/5.1/manual.html#2.1
+local keywords = {}
+for _, keyword in pairs({
+	"and", "break", "do", "else", "elseif",
+	"end", "false", "for", "function", "if",
+	"in", "local", "nil", "not", "or",
+	"repeat", "return", "then", "true", "until", "while"
+}) do
+	keywords[keyword] = true
+end
+
+-- Build a table with the succeeding character from A-Z
+-- Used to increment references quickly
+local succ_ref_char = {}
+for char = ("A"):byte(), ("Z"):byte() - 1 do
+	succ_ref_char[string.char(char)] = string.char(char + 1)
+end
 
-	-- Mutually recursive local functions:
-	local dump_val, dump_or_ref_val
+local function quote(string)
+	return string_format("%q", string)
+end
+
+local function dump_func(func)
+	return string_format("loadstring(%q)", string_dump(func))
+end
 
-	-- If x occurs multiple times, dump the local variable rather than
-	-- the value. If it's the first time it's dumped, also dump the
-	-- content in local_defs.
-	function dump_or_ref_val(x)
-		if seen[x] ~= 2 then
-			return dump_val(x)
+-- Serializes Lua nil, booleans, numbers, strings, tables and even functions
+-- Tables are referenced by reference, strings are referenced by value. Supports circular tables.
+local function serialize(value, write)
+	-- References are stored as tables of characters A-Z, which are digits of a base 26 number
+	local reference = {"A"}
+	local function increment_reference(place)
+		if not reference[place] then
+			reference[place] = "B"
+		elseif reference[place] == "Z" then
+			-- Carriage
+			reference[place] = "A"
+			return increment_reference(place + 1)
+		else
+			reference[place] = succ_ref_char[reference[place]]
 		end
-		local var = dumped[x]
-		if var then  -- Already referenced
-			return var
+	end
```

Comment:
I have the same thought here: why not just use `"r" .. n`?
The amount of saved space is minimal and keeping a table around that you have to increment and then concat (this was not immediately obvious to me btw) is IMO wasted work.

Reply:
> The amount of saved space is minimal

It is a factor of `math.log(26)/math.log(10)` which is ca. `1.4` plus the one wasted character `r`. This probably makes for 2-3 characters per ref; on the other hand, not stringification of numbers might be faster. Can I at least use hex :>?

Reply:
Sure

Reply:
Done